rust/otap-dataflow/crates/telemetry/src/instrument.rs

```diff
@@ -0,0 +1,500 @@
+    fn from(value: u64) -> Self {
+        UpDownCounter(value)
+    }
+}
+
+impl AddAssign<u64> for UpDownCounter<u64> {
+    fn add_assign(&mut self, rhs: u64) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: UpDownCounter arithmetic is expected to be well-behaved in telemetry scenarios.
+            // Wrapping behavior is acceptable for performance-critical metric operations.
+            self.0 = self.0.wrapping_add(rhs);
+        }
+        #[cfg(not(feature = "unchecked-arithmetic"))]
+        {
+            self.0 += rhs;
+        }
+    }
+}
+
+impl SubAssign<u64> for UpDownCounter<u64> {
+    fn sub_assign(&mut self, rhs: u64) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: UpDownCounter subtraction is expected to be well-behaved in telemetry scenarios.
+            // Wrapping behavior is acceptable for performance-critical metric operations.
+            self.0 = self.0.wrapping_sub(rhs);
+        }
+        #[cfg(not(feature = "unchecked-arithmetic"))]
+        {
+            self.0 -= rhs;
+        }
+    }
+}
+
+impl UpDownCounter<u64> {
+    /// Increments the up-down-counter by 1.
+    #[inline]
+    pub fn inc(&mut self) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: Incrementing by 1 is safe for wrapping arithmetic in telemetry contexts
+            self.0 = self.0.wrapping_add(1);
+        }
+        #[cfg(not(feature = "unchecked-arithmetic"))]
+        {
+            self.0 += 1;
+        }
+    }
+
+    /// Decrements the up-down-counter by 1.
+    #[inline]
+    pub fn dec(&mut self) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: Decrementing by 1 is safe for wrapping arithmetic in telemetry contexts
+            self.0 = self.0.wrapping_sub(1);
+        }
+        #[cfg(not(feature = "unchecked-arithmetic"))]
+        {
+            self.0 -= 1;
+        }
+    }
+}
+
+// Gauge implementation.
+// =====================
+
+impl<T> Gauge<T>
+where
+    T: Copy
+        + Default
+        + std::ops::Add<Output = T>
+        + std::ops::Sub<Output = T>
+        + AddAssign
+        + SubAssign,
+{
+    /// Creates a new gauge initialized with the provided value.
+    #[inline]
+    pub const fn new(v: T) -> Self {
+        Self(v)
+    }
+
+    /// Reset the gauge to 0.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.0 = T::default();
+    }
+
+    /// Sets the value of the gauge.
+    #[inline]
+    pub fn set(&mut self, v: T) {
+        self.0 = v;
+    }
+
+    /// Returns the current value of the gauge.
+    #[inline]
+    pub fn get(&self) -> T {
+        self.0
+    }
+
+    /// Adds an arbitrary value to the gauge.
+    #[inline]
+    pub fn add(&mut self, v: T) {
+        self.0 += v;
+    }
+
+    /// Subs an arbitrary value to the gauge.
+    #[inline]
+    pub fn sub(&mut self, v: T) {
+        self.0 -= v;
+    }
```

Comment:
```suggestion
```
In a framework where UpDownCounter exists, the gauge instrument shouldn't expose add/sub. If you can add / subtract from it, it's not a gauge (IMO).

Reply:
Good point. I will remove the add/subtract methods as well as the AddAssign and SubAssign.